File: test/neutron_star.jl
# Distributed under the MIT License.
# See LICENSE for details.

using LinearAlgebra
using Random
using StaticArrays
using Test

include("../src/neutron_star.jl")

function test()
    seed = Random.rand(1:(10^10))
    rng = Xoshiro(seed)

    @testset verbose = true "NeutronStar | Seed: $seed" begin
        @testset "Class" begin
            mass = Random.rand(rng)
            radius = Random.rand(rng)
            period = Random.rand(rng)
            unn_rotation_axis = SVector{3, Float64}(Random.randn(rng, 3))
            magnetic_moment = SVector{3, Float64}(Random.randn(rng, 3))
            ns = NeutronStar(
                mass,
                radius,
                period,
                unn_rotation_axis,
                magnetic_moment
            )

            @test isa(ns, NeutronStar)
            @test isapprox(mass, ns.mass)
            @test isapprox(radius, ns.radius)
            @test isapprox(period, ns.period)
            @test isapprox(
                unn_rotation_axis / norm(unn_rotation_axis),
                ns.rotation_axis
            )
            @test isapprox(magnetic_moment, ns.magnetic_moment)
            @test isapprox((2.0 * pi / period), norm(ns.angular_velocity))
        end

        @testset "ConversionRadius" begin
            n_the::UInt64 = Random.rand(rng, 2:5)
            angle = Random.randn(rng, n_the)
            rad = Vector{Float64}(undef, n_the)
            conversion_radius!(rad, angle)

            rad_expected =
                abs.((3.0 * cos.(angle) .^ 2.0 .- 1.0) / 2.0) .^ (1.0 / 3.0)

            @test isapprox(rad, rad_expected)
            @test_throws AssertionError conversion_radius!(
                Random.randn(rng, 3),
                Random.randn(rng, 4)
            )
        end

        @testset "ConversionSurface" begin
            n_the = Random.rand(rng, 2:5)
            n_phi::UInt64 = Random.rand(rng, 2:5)
            n_surf = n_the * n_phi

            theta = Random.randn(rng, n_the)
            phi = Random.randn(rng, n_phi)
            surf = Matrix{Float64}(undef, n_surf, 3)
            conversion_surface!(surf, theta, phi)

            rad = Random.randn(rng, size(theta))
            conversion_radius!(rad, theta)

            surf_expected = Matrix{Float64}(undef, n_surf, 3)
            for k in 1:n_phi
                inc = (k - 1) * n_the
                for j in 1:n_the
                    q = j + inc
                    surf_expected[q, 1] = rad[j] * sin(theta[j]) * cos(phi[k])
                    surf_expected[q, 2] = rad[j] * sin(theta[j]) * sin(phi[k])
                    surf_expected[q, 3] = rad[j] * cos(theta[j])
                end
            end

            @test isapprox(surf, surf_expected)
            @test_throws AssertionError conversion_surface!(
                Random.randn(rng, 3, 4),
                Random.randn(rng, 2),
                Random.randn(rng, 3)
            )
        end
    end
    return nothing
end

test()

File: post/neutron_star.py
1  # Distributed under the MIT License.
# See LICENSE for details.

import numpy as np


class NeutronStar:

    def __init__(self, mass, radius, rotation_period, rotation_axis, b_zero,
                 magnetic_axis):
        self._mass = mass
        self._radius = radius
        self._rotation_period = rotation_period
        self._rotation_axis = rotation_axis / np.linalg.norm(rotation_axis)
        self._b_zero = b_zero
        self._magnetic_axis = magnetic_axis / np.linalg.norm(magnetic_axis)

        self._angular_speed = 2. * np.pi / self._rotation_period
        self._angular_velocity = self._angular_speed * self._rotation_axis
        self._phase = np.arctan2(magnetic_axis[1], magnetic_axis[0])

    @property
    def mass(self):
        return self._mass

    @property
    def radius(self):
        return self._radius

    @property
    def rotation_period(self):
        return self._rotation_period

    @property
    def rotation_axis(self):
        return self._rotation_axis

    @property
    def angular_speed(self):
        return self._angular_speed

    @property
    def angular_velocity(self):
        return self._angular_velocity

    @property
    def phase(self):
        return self._phase

    @property
    def b_zero(self):
        return self._b_zero

    @property
    def magnetic_axis(self):
        return self._magnetic_axis
